File: core/core.net/source/core.net/Utility/BitConverter/BitConverter.cs
//
// Copyright (c) 2015. SeongJun Park.
//
// Distributed under the MIT License.
//

using System;

namespace Useless.Utility
{
	public class BitConverter
	{
		public static byte[] GetBytes( Type type, object val )
		{
			if( type == typeof( bool ) )
			{
				return System.BitConverter.GetBytes( ( bool )val );
			}
			else if( type == typeof( sbyte ) )
			{
				return System.BitConverter.GetBytes( ( sbyte )val );
			}
			else if( type == typeof( byte ) )
			{
				return System.BitConverter.GetBytes( ( byte )val );
			}
			else if( type == typeof( short ) )
			{
				return System.BitConverter.GetBytes( ( short )val );
			}
			else if( type == typeof( ushort ) )
			{
				return System.BitConverter.GetBytes( ( ushort )val );
			}
			else if( type == typeof( int ) )
			{
				return System.BitConverter.GetBytes( ( int )val );
			}
			else if( type == typeof( uint ) )
			{
				return System.BitConverter.GetBytes( ( uint )val );
			}
			else if( type == typeof( long ) )
			{
				return System.BitConverter.GetBytes( ( long )val );
			}
			else if( type == typeof( ulong ) )
			{
				return System.BitConverter.GetBytes( ( ulong )val );
			}
			else if( type == typeof( float ) )
			{
				return System.BitConverter.GetBytes( ( float )val );
			}
			else if( type == typeof( double ) )
			{
				return System.BitConverter.GetBytes( ( double )val );
			}
			else if( type == typeof( char ) )
			{
				return System.BitConverter.GetBytes( ( char )val );
			}

			throw new NotSupportedException();
			
		}

		public static bool ToBoolean( byte[] bytes )
		{
			return System.BitConverter.ToBoolean( bytes, 0 );
		}

		public static char ToChar( byte[] bytes )
		{
			return System.BitConverter.ToChar( bytes, 0 );
		}

		public static sbyte ToS8( byte[] bytes )
		{
			return ( sbyte )bytes[ 0 ];
		}

		public static byte ToU8( byte[] bytes )
		{
			return bytes[ 0 ];
		}

		public static short ToS16( byte[] bytes )
		{
			return System.BitConverter.ToInt16( bytes, 0 );
		}

		public static ushort ToU16( byte[] bytes )
		{
			return System.BitConverter.ToUInt16( bytes, 0 );
		}

		public static int ToS32( byte[] bytes )
		{
			return System.BitConverter.ToInt32( bytes, 0 );
		}

		public static uint ToU32( byte[] bytes )
		{
			return System.BitConverter.ToUInt32( bytes, 0 );
		}

		public static long ToS64( byte[] bytes )
		{
			return System.BitConverter.ToInt64( bytes, 0 );
		}

		public static ulong ToU64( byte[] bytes )
		{
			return System.BitConverter.ToUInt64( bytes, 0 );
		}

		public static float ToF32( byte[] bytes )
		{
			return System.BitConverter.ToSingle( bytes, 0 );
		}

		public static double ToF64( byte[] bytes )
		{
			return System.BitConverter.ToDouble( bytes, 0 );
		}
		
		public static object To( Type type, byte[] bytes )
		{
			if( type == typeof( bool ) )
			{
				return ToBoolean( bytes );
			}
			else if( type == typeof( sbyte ) )
			{
				return ToS8( bytes );
			}
			else if( type == typeof( byte ) )
			{
				return ToU8( bytes );
			}
			else if( type == typeof( short ) )
			{
				return ToS16( bytes );
			}
			else if( type == typeof( ushort ) )
			{
				return ToU16( bytes );
			}
			else if( type == typeof( int ) )
			{
				return ToS32( bytes );
			}
			else if( type == typeof( uint ) )
			{
				return ToU32( bytes );
			}
			else if( type == typeof( long ) )
			{
				return ToS64( bytes );
			}
			else if( type == typeof( ulong ) )
			{
				return ToU64( bytes );
			}
			else if( type == typeof( float ) )
			{
				return ToF32( bytes );
			}
			else if( type == typeof( double ) )
			{
				return ToF64( bytes );
			}
			else if( type == typeof( char ) )
			{
				return ToChar( bytes );
			}

			throw new NotSupportedException();
		}

		public static T To<T>( byte[] bytes )
		{
			return ( T )( To( typeof( T ), bytes ) );
		}
	}
}

File: core/core.java/source/core.java/utility/src/useless/utility/BitConverter.java
/*
 * Copyright (c) 2015. SeongJun Park.
 *
 * Distributed under the MIT License.
 */

package useless.utility;

import com.sun.org.apache.xpath.internal.operations.Bool;

public class BitConverter
{
	public static byte[] getBytes( boolean value )
	{
		return new byte[]
				{
					( byte )( value ? 1 : 0 )
				};
	}

	public static byte[] getBytes( byte value )
	{
		return new byte[]
				{
					value
				};
	}

	public static byte[] getBytes( char value )
	{
		return new byte[]
				{
					( byte )( ( value ) & 0xFF ),
					( byte )( ( value >> 8 ) & 0xFF )
				};
	}

	public  static byte[] getBytes( short value )
	{
		return new byte[]
				{
					( byte ) ( ( value ) & 0xFF ),
					( byte ) ( ( value >> 8 ) & 0xFF )
				};
	}

	public static byte[] getBytes( int value )
	{
		return new byte[]
				{
					( byte )( ( value ) & 0xFF ),
					( byte )( ( value >> 8 ) & 0xFF ),
					( byte )( ( value >> 16 ) & 0xFF ),
					( byte )( ( value >> 24 ) & 0xFF )
				};
	}

	public static byte[] getBytes( long value )
	{
		return new byte[]
				{
					( byte )( ( value ) & 0xFF ),
					( byte )( ( value >> 8 ) & 0xFF ),
					( byte )( ( value >> 16 ) & 0xFF ),
					( byte )( ( value >> 24 ) & 0xFF ),
					( byte )( ( value >> 32 ) & 0xFF ),
					( byte )( ( value >> 40 ) & 0xFF ),
					( byte )( ( value >> 48 ) & 0xFF ),
					( byte )( ( value >> 56 ) & 0xFF ),
				};
	}

	public static byte[] getBytes( float value )
	{
		return getBytes( Float.floatToRawIntBits( value ) );
	}

	public static byte[] getBytes( double value )
	{
		return getBytes( Double.doubleToRawLongBits( value ) );
	}

	public static byte[] getBytes( String value )
	{
		return value.getBytes();
	}

	public static byte[] getBytes_U8( short value )
	{
		return new byte[]
				{
						( byte ) ( ( value ) & 0xFF )
				};
	}

	public static byte[] getBytes_U16( int value )
	{
		return new byte[]
				{
						( byte ) ( ( value ) & 0xFF ),
						( byte ) ( ( value >> 8 ) & 0xFF )
				};
	}

	public static byte[] getBytes_U32( long value )
	{
		return new byte[]
				{
						( byte )( ( value ) & 0xFF ),
						( byte )( ( value >> 8 ) & 0xFF ),
						( byte )( ( value >> 16 ) & 0xFF ),
						( byte )( ( value >> 24 ) & 0xFF )
				};
	}

	public static <T> byte[] getBytes( Class<T> type, T value )
	{
		if( type == Boolean.class || type == boolean.class )
		{
			return getBytes( ( Boolean )value );
		}
		else if( type == Character.class || type == char.class )
		{
			return getBytes( ( Character )value );
		}
		else if( type == Byte.class || type == byte.class )
		{
			return getBytes( ( Byte )value );
		}
		else if( type == Short.class || type == short.class )
		{
			return getBytes( ( Short )value );
		}
		else if( type == Integer.class || type == int.class )
		{
			return getBytes( ( Integer )value );
		}
		else if( type == Long.class || type == long.class )
		{
			return getBytes( ( Long )value );
		}
		else if( type == Float.class || type == float.class )
		{
			return getBytes( ( Float )value );
		}
		else if( type == Double.class || type == double.class )
		{
			return getBytes( ( Double )value );
		}

		throw new UnsupportedOperationException();
	}

	public static boolean toBoolean( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 1 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( bytes[ 0 ] != 0 );
	}

	public static char toChar( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 2 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( char )
				(
					( ( char )( bytes[ 0 ] ) << 8 ) |
					( ( char )( bytes[ 1 ] ) )
				);
	}

	public static byte toS8( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 1 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return bytes[ 0 ];
	}

	public static short toU8( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 1 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( short )bytes[ 0 ];
	}

	public static short toS16( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 2 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( short )
				(
					( ( short )( bytes[ 0 ] ) |
					( ( short )( bytes[ 1 ] ) << 8 ) )
				);
	}

	public static int toU16( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 2 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( int )
				(
					( ( short )( bytes[ 0 ] ) |
					( ( short )( bytes[ 1 ] ) << 8 ) )
				);
	}

	public static int toS32( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 4 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( int )
				(
					( ( int )( bytes[ 0 ] ) |
					( ( int )( bytes[ 1 ] ) << 8 ) |
					( ( int )( bytes[ 2 ] ) << 16 ) |
					( ( int )( bytes[ 3 ] ) << 24 ) )
				);
	}

	public static long toU32( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 4 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( long )
				(
					( ( long )( bytes[ 0 ] ) |
					( ( long )( bytes[ 1 ] ) << 8 ) |
					( ( long )( bytes[ 2 ] ) << 16 ) |
					( ( long )( bytes[ 3 ] ) << 24 ) )
				);
	}

	public static long toS64( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 8 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return ( long )
				(
					( ( long )( bytes[ 0 ] ) |
					( ( long )( bytes[ 1 ] ) << 8 ) |
					( ( long )( bytes[ 2 ] ) << 16 ) |
					( ( long )( bytes[ 3 ] ) << 24 ) |
					( ( long )( bytes[ 4 ] ) << 32 ) |
					( ( long )( bytes[ 5 ] ) << 40 ) |
					( ( long )( bytes[ 6 ] ) << 48 ) |
					( ( long )( bytes[ 7 ] ) << 56 ) )
				);
	}

	public static float toF32( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 4 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return Float.intBitsToFloat( toS32( bytes ) );
	}

	public static double toF64( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null || bytes.length < 8 )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return Double.longBitsToDouble( toS64( bytes ) );
	}

	public static String toString( byte[] bytes ) throws IllegalArgumentException
	{
		if( bytes == null )
		{
			throw new IllegalArgumentException( "bytes" );
		}

		return new String( bytes );
	}

	public static <T> T to( Class<T> type, byte[] bytes )
	{
		if( type == Boolean.class || type == boolean.class )
		{
			return ( T )( Object )toBoolean( bytes );
		}
		else if( type == Character.class || type == char.class )
		{
			return ( T )( Object )toChar( bytes );
		}
		else if( type == Byte.class || type == byte.class )
		{
			return ( T )( Object )toS8( bytes );
		}
		else if( type == Short.class || type == short.class )
		{
			return ( T )( Object )toS16( bytes );
		}
		else if( type == Integer.class || type == int.class )
		{
			return ( T )( Object )toS32( bytes );
		}
		else if( type == Long.class || type == long.class )
		{
			return ( T )( Object )toS64( bytes );
		}
		else if( type == Float.class || type == float.class )
		{
			return ( T )( Object )toF32( bytes );
		}
		else if( type == Double.class || type == double.class )
		{
			return ( T )( Object )toF64( bytes );
		}

		throw new UnsupportedOperationException();
	}
}
